imap_processing/ena_maps/ena_maps.py

```diff
@@ -867,6 +870,304 @@ def __init__(
+                    + absolute_tolerance
+                )
+                if abs_delta < total_abs_tolerance:
+                    break
+            depth += 1
+
+        # Only keep the last (best) mean pixel value
+        return mean_pixel_value_at_level[-1], depth
+
+    def to_rectangular_skymap(
+        self,
+        rect_spacing_deg: float,
+        value_keys: list[str],
+        max_subdivision_depth: int = MAX_SUBDIV_RECURSION_DEPTH,
+    ) -> tuple[RectangularSkyMap, dict[str, np.typing.NDArray]]:
+        """
+        Interpolate a healpix map to a rectangular map using recursive subdivision.
+
+        Parameters
+        ----------
+        rect_spacing_deg : float
+            The spacing of the rectangular map in degrees.
+        value_keys : list[str]
+            The names of the values to interpolate from the healpix map.
+            Each must be independently interpolated because the subdivision depth
+            depends on the gradient of the value between adjacent healpix pixels.
+        max_subdivision_depth : int, optional
+            The maximum depth of recursion for subdivision,
+            by default MAX_SUBDIV_RECURSION_DEPTH.
+
+        Returns
+        -------
+        tuple[RectangularSkyMap, dict[str, np.typing.NDArray]]
+            A RectangularSkyMap containing the interpolated values, and a dictionary of
+            each value and its corresponding subdivision depth by pixel.
+        """
+        # Begin by defining the rectangular map we want to create, which must be
+        # in the same spice reference frame as the healpix map
+        rect_map = RectangularSkyMap(
+            spacing_deg=rect_spacing_deg,
+            spice_frame=self.spice_reference_frame,
+        )
+
+        # Depending on the maximum recursion depth, the number of pixels in the
+        # RectangularSkyMap, and the number of value keys, and especially on the
+        # gradients of the values, the number of operations can be very large, so
+        # log key information about the expected number of operations.
+        approx_max_operations = (
+            (4**max_subdivision_depth) * self.num_points * len(value_keys)
+        )
+        logger.info(
+            f"Converting from a HealpixSkyMap(nside={self.nside}) to a "
+            f"RectangularSkyMap(spacing_deg={rect_spacing_deg}) with recursive "
+            "subdivision.\n The maximum recursion depth is "
+            f"{max_subdivision_depth}, yielding a maximum number of healpix calls"
+            f" of {approx_max_operations:.3e}."
+        )
+
+        # Dict to hold the subdivision depth by pixel for each value key
+        subdiv_depth_dict = {}
+        for value_key in value_keys:
+            self.data_1d[value_key]
+
+            # For each of the values, calculate each pixel's value with
+            # recursive subdivision. Unfortunately, this must be done independently
+            # for each value key.
```

Comment:
Depending on the number of values covering the map I assume? I am kind of curious if it is worthwhile just making the maximum recursion depth mapping: 256 x 256 x nside once (~8M) points, and then re-use that mapping for every single value as a kind of index lookup table. So just live with the fact that you'll do one expensive calculation of subdivisions, but use that knowledge elsewhere in your interpolations to speed those subsequent steps up.